Static linking error, missing some link for curand/cublas/cublaslt

Command:
```
cargo test --tests --no-default-features -F cuda-version-from-build-system,std,curand,static-linking
```

nvidia-smi output:
```
Tue Apr  8 15:17:00 2025       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 570.124.06             Driver Version: 570.124.06     CUDA Version: 12.8     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA A10                     On  |   00000000:06:00.0 Off |                    0 |
|  0%   28C    P8             15W /  150W |       1MiB /  23028MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
```

See attached file for full output.



[error.txt](https://github.com/user-attachments/files/19651997/error.txt)

Selected snippets:

```
...
usr/bin/ld: curand_mt19937_functions.compute_120.cudafe1.cpp:(.text+0x8474): undefined reference to `__cudaPushCallConfiguration'
          /usr/bin/ld: curand_mt19937_functions.compute_120.cudafe1.cpp:(.text+0x8606): undefined reference to `__cudaPopCallConfiguration'
          /usr/bin/ld: curand_mt19937_functions.compute_120.cudafe1.cpp:(.text+0x8650): undefined reference to `cudaLaunchKernel'
/usr/bin/ld: curand_mt19937_functions.compute_120.cudafe1.cpp:(.text+0x3d53): undefined reference to `cudaFree'
/usr/bin/ld: curand_mt19937_functions.compute_120.cudafe1.cpp:(.text+0x3a16): undefined reference to `cudaMemcpy'
...
```

--- a/build.rs
+++ b/build.rs
@@ -199,36 +199,30 @@ fn static_linking(major: usize, minor: usize) {
         }
     }
 
-    println!("cargo:rustc-link-lib=dylib=stdc++");
-    #[cfg(feature = "driver")]
-    println!("cargo:rustc-link-lib=dylib=cuda");
+    println!("cargo:rustc-link-lib=static:+whole-archive=stdc++");
+    #[cfg(any(feature = "driver", feature = "runtime"))]
+    {
+        println!("cargo:rustc-link-lib=dylib=cuda");
+        println!("cargo:rustc-link-lib=static:+whole-archive=cudart_static");
+    }
     #[cfg(feature = "nccl")]
-    println!("cargo:rustc-link-lib=dylib=nccl");
-    #[cfg(feature = "runtime")]
-    println!("cargo:rustc-link-lib=static=cudart_static");
+    println!("cargo:rustc-link-lib=static:+whole-archive=nccl_static");
     #[cfg(feature = "nvrtc")]
     {
-        println!("cargo:rustc-link-lib=static=nvrtc_static");
-        println!("cargo:rustc-link-lib=static=nvptxcompiler_static");
-        println!("cargo:rustc-link-lib=static=nvrtc-builtins_static");
+        println!("cargo:rustc-link-lib=static:+whole-archive=nvrtc_static");
+        println!("cargo:rustc-link-lib=static:+whole-archive=nvptxcompiler_static");
+        println!("cargo:rustc-link-lib=static:+whole-archive=nvrtc-builtins_static");
     }
+    #[cfg(any(feature = "curand", feature = "cublas", feature = "cublaslt"))]
+    println!("cargo:rustc-link-lib=static:+whole-archive=culibos");
     #[cfg(feature = "curand")]
-    {
-        println!("cargo:rustc-link-lib=static=culibos");
-        println!("cargo:rustc-link-lib=static=curand_static");
-    }
+    println!("cargo:rustc-link-lib=static:+whole-archive=curand_static");
     #[cfg(feature = "cublas")]
-    {
-        println!("cargo:rustc-link-lib=static=culibos");
-        println!("cargo:rustc-link-lib=static=cublas_static");
-    }
-    #[cfg(any(feature = "cublas", feature = "cublaslt"))]
-    {
-        println!("cargo:rustc-link-lib=static=culibos");
-        println!("cargo:rustc-link-lib=static=cublasLt_static");
-    }
+    println!("cargo:rustc-link-lib=static:+whole-archive=cublas_static");
+    #[cfg(feature = "cublaslt")]
+    println!("cargo:rustc-link-lib=static:+whole-archive=cublasLt_static");
     #[cfg(feature = "cudnn")]
-    println!("cargo:rustc-link-lib=static=cudnn");
+    println!("cargo:rustc-link-lib=static:+whole-archive=cudnn");
 }
 
 #[allow(unused)]

